vermouth/tests/test_name_moltype.py

```diff
@@ -0,0 +1,151 @@
+    # their order.
+    n_moltypes = draw(st.integers(min_value=1, max_value=max_moltypes))
+    unique_pre_moltypes = range(n_moltypes)
+    pre_moltypes_sampled = draw(st.lists(st.sampled_from(unique_pre_moltypes),
+                                         min_size=min_size, max_size=max_size))
+    if not pre_moltypes_sampled:
+        return [], []
+
+    moltype_index = 0
+    pre_to_moltype = {pre_moltypes_sampled[0]: 'molecule_{}'.format(moltype_index)}
+    moltypes = []
+    for pre_moltype in pre_moltypes_sampled:
+        if pre_moltype not in pre_to_moltype:
+            moltype_index += 1
+            pre_to_moltype[pre_moltype] = 'molecule_{}'.format(moltype_index)
+        moltypes.append(pre_to_moltype[pre_moltype])
+
+    sharing = defaultdict(lambda : draw(random_molecule(**molecule_kwargs)))
+    molecules = [sharing[moltype] for moltype in moltypes]
+
+    # So far we made sure that molecules with a different moltype are drawn
+    # separately. Though, two different draws may result in equal molecules,
+    # especially as molecules are reduced whn shrinking the draws.
+    representatives = [
+        next(group)[0]
+        for _, group in itertools.groupby(
+                zip(molecules, moltypes), key=lambda x: x[1])
+    ]
+    assume(all(not mol1.share_moltype_with(mol2)
+               for mol1, mol2 in itertools.combinations(representatives, 2)))
+
+    # Each molecule must be its own instance as molecules are modified in place.
+    molecules = [molecule.copy() for molecule in molecules]
+
+    return [molecules, moltypes]
+
+
+@pytest.mark.parametrize('deduplicate', (True, False))
+@given(mols_and_moltypes=molecules_and_moltypes(
+    max_size=4, max_moltypes=3,  molecule_kwargs={'max_nodes': 3, 'max_meta': 3},
+))
+def test_name_moltype(mols_and_moltypes, deduplicate):
+    molecules, moltypes = mols_and_moltypes
+
+    system = System()
+    system.molecules = molecules
+    
+    if not deduplicate:
+        moltypes = ['molecule_{}'.format(i) for i in range(len(molecules))]
+
+    processor = NameMolType(deduplicate=deduplicate)
+    processor.run_system(system)
+
+    found_moltypes = [molecule.meta['moltype'] for molecule in system.molecules]
+    assert found_moltypes == moltypes
+
+
+@pytest.mark.parametrize('deduplicate', (True, False))
+def test_martinize2_moltypes(tmpdir, deduplicate):
+    """
+    Run martinize2 and make sure the ITP file produced have the expected names.
+    """
+    command = [
+        'martinize2',
+        '-f', str(PDB_HB),
+        '-o', 'topol.top',
+        '-x', 'out.pdb',
+        '-ignore', 'HOH', '-ignore', 'HEME',
+    ]
+    if deduplicate:
+        expected = ['molecule_{}.itp'.format(i) for i in range(2)]
+        n_outputs = 2
+    else:
+        command.append('-sep')
+        n_outputs = 4
+    expected = ['molecule_{}.itp'.format(i) for i in range(n_outputs)]
+
+    proc = subprocess.Popen(command, cwd=str(tmpdir))
+    exit_code = proc.wait()
```

Comment:
Is there a chance this will fail if proc produces a lot of stdout/stderr output?
And maybe set a timeout